Feature/user controller user dto

xenex90 opened the pull request:
@f-lab-lyan 멘토님

컨트롤러, 테스트, 도커 컴포즈 설정 파일
push 하였습니다.

기능적인 부분은 없이 간단히 구현하여 올렸습니다.
확인 부탁드리겠습니다.

xenex90 committed:
멀티모듈 프로젝트 구조 전환

- 모놀리틸에서 멀티모듈 아키텍처로 변경
- user-service: 사용자 관리 및 인증 기능
- content-service: 콘텐츠 생성 및 관리 기능
- community-service: 커뮤니티 기능 및 상호작용
- code-review-service: 코드 리뷰 기능
- common: 공통 유틸리티 및 DTO 모듈
- database: 엔티티 및 리포지토리 계층
- Gradle 멀티모듈 빌드 시스템 구성
- Spring Boot 3.5.0, Java 21 환경 설정"

xenex90 committed:
chore: 프로젝트 구조 리팩토링 및 Gradle convention 플러그인 적용

- 기존 모듈별/루트 Gradle 설정(build.gradle.kts, settings.gradle.kts, wrapper 등)과 application.properties 삭제
- buildSrc에 convention-java, convention-spring-boot 플러그인 추가
- 각 서브모듈은 convention 플러그인만 적용하도록 build.gradle.kts 재구성
- 공통 설정(자바 툴체인, 리포지토리, BOM 등)을 중앙화하여 중복 제거

xenex90 committed:
불필요한 주석 제거

xenex90 committed:
Delete everybodycodeacademy/gradlew.bat

xenex90 committed:
Delete everybodycodeacademy/gradle/wrapper/gradle-wrapper.properties

불필요한 파일 제거

xenex90 committed:
Delete everybodycodeacademy/gradlew

불필요한 파일 제거

xenex90 committed:
Add userConteroller, unitTest, IntegrationTest, docker-compose.yml

-회원가입, 로그인, 로그아웃 Contoller 틀 추가
-UserControllerInegrationTest, UserControllerUnitTest 추가
-도커 컴포즈 (DB) 파일 추가

xenex90 committed:
Merge branch 'feature/user_controller_userDTO' of https://github.com/f-lab-edu/everybodycodeacademy into feature/user_controller_userDTO

xenex90 commented on user-service/src/main/java/flab/userservice/controller/userController.java:
```diff
@@ -0,0 +1,49 @@
+    @PostMapping("/signup")
+    @ResponseStatus(HttpStatus.CREATED)
+    @ResponseBody
+    public String signup(@RequestBody UserSignupRequest request) {
```
기능적인 부분은 없이 틀만 올렸습니다.

xenex90 commented on user-service/src/test/java/flab/userservice/UserControllerIntegrationTest.java:
```diff
@@ -0,0 +1,67 @@
+
+    @Test
+    @DisplayName("회원가입 성공 시 201 반환")
+    void signupSuccess() throws Exception {
```
동작하는 지는 확인 안해본 상태입니다.
추후 해보겠습니다.

xenex90 commented on docker-compose.yml:
```diff
@@ -0,0 +1,19 @@
+      MYSQL_PASSWORD: test4321
+    ports:
+      - "3306:3306"                        # 호스트:컨테이너 포트 매핑
+    volumes:
```
도커 컴포즈로 도커 DB 생성되며 동작 하는 것 까지 확인하엿습니다.

xenex90 reviewed (commented):
파일 변경사항 확인하였습니다.

xenex90 committed:
-userDTO modified, enum loginProvider add

- UserSignupRequest: 회원가입 요청 DTO 수정
  - 필수 필드: username, email, password, phoneNumber 등
  - Validation 추가: 최소/최대 길이, 이메일 형식, 전화번호 패턴, 과거 날짜 등
- UserLoginRequest: 로그인 요청 DTO 수정
  - LOCAL 및 소셜 로그인 구분을 위한 loginProvider 필드 포함
  - Validation 추가: 로그인 제공자를 필수로 지정
- UserResponse: 사용자 응답 DTO 수정
  - 생성일자, 마지막 로그인 시간 포함
  - 사용자 데이터 응답을 위한 필드 정의

xenex90 commented on docker-compose.yml:
```diff
@@ -0,0 +1,19 @@
+    environment:
+      MYSQL_ROOT_PASSWORD: kyjtester
+      MYSQL_DATABASE: everycodeacademy
+      MYSQL_USER: kyjtester1
```
멘토님, docker-compose.yml 파일은 깃허브에 올리면 보안상 위험하지 않나요?

f-lab-lyan replied:
이 도커 컴포즈를 어떤 용도로 사용할 것인지에 따라 다르지 않을까요? 일반적으로 local에서 테스트하는 용도라면, 전혀 문제없습니다.

xenex90 commented on user-service/src/main/java/flab/userservice/dto/userDTO.java:
```diff
@@ -0,0 +1,62 @@
+import jakarta.validation.constraints.Email;
+import jakarta.validation.constraints.NotBlank;
+
+public record UserSignupRequest(
```
몇일 전 commit에서 수정하였습니다.
userDTO의 정석 코드입니다.

xenex90 commented on user-service/src/main/java/flab/userservice/dto/loginProvider.java:
```diff
@@ -0,0 +1,3 @@
+public enum LoginProvider {
+    LOCAL, KAKAO, GOOGLE, NAVER;
```
유지보수성 고려해서 enum으로 추가했습니다.

f-lab-lyan replied:
이걸 나중에 어떻게 관리하실 생각이신가요? 각각 설정이 다 다르지 않을까요?

xenex90 replied:
기본 값은 Local로 고정하고, 나머지는 Social Login Provider는 추후 상황을 봐서 사용할 예정입니다.
SNS login 기능 구현은 아직 전체 과정을 다 구현해보지 않아서, 이번에 할 수 있으면 할려고 미리 넣어두었습니다.
취업 시장에서 SNS login 제법 자주 사용 하는 기능인 것 같아서 해보고 싶었습니다.

각 Social Login Provider 마다 구현 코드가 다르지만, 기본적인 id, password 를 입력해서 인증, 인가를 거치면 회원 고유값(ci)을 response로 반환해주는 것 까진 알고 있습니다.

xenex90 reviewed (commented):
userDTO 수정 후 확인했습니다!
리뷰 부탁드립니다 ㅎ

f-lab-lyan commented on user-service/src/main/java/flab/userservice/controller/userController.java:
```diff
@@ -0,0 +1,49 @@
+    @ResponseBody
+    public String signup(@RequestBody UserSignupRequest request) {
+        //userService.signup(request);
+        return "CREATED";
```
리턴 밸류로 스트링이 아니라 다른 걸 줘야하지 않을까요? 예를 들면 생성된 유저정보라던가 ...

xenex90 replied:
아 깜박했네요 수정하겠습니다.

f-lab-lyan commented on user-service/src/test/java/flab/userservice/UserControllerIntegrationTest.java:
```diff
@@ -0,0 +1,67 @@
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+@SpringBootTest
+@AutoConfigureMockMvc
```
이것 대신에 @SpringBootTest(...RANDOM_PORT)를 써서 테스트를 작성해 보시면 좋을 것 같아요. 그리고, 그 두개는 테스트의 목적에서 어떤 차이가 있을까요?

f-lab-lyan replied:
그리고 Request와 Response를 둘 다 테스트 해보시면 좋을 것 같아요.

xenex90 replied:
테스트는 컨트롤러 기능이 어느 정도 있어야 되지 않을까요?
현재코드 기준으로 딱히 동작 테스트 할 건 없어 보이기도 하고, 웹 브라우저에 해당 컨트롤러 endpoint 입력해도, String return값만 뜰 것입니다.

멘토님, 또 의문인게 있습니다.
Integration Test + Unit Test을 올리라고 하셔서 고민을 하다가, 멘토님 아직 컨트롤러가 완성안된상태에서 test코드가 완성될리가 없는데, 부분부분 비즈니스 로직 개발 과정이나 완성된 이후에 test를 하기 위한 것이 unitTest라고 알고 있습니다.

제가 모르는 것이 있거나 잘 못 알 고 있을 순 있어서 우선
업무에서는 이렇게 하면 큰일 나겠지만
간단한 코드 올리고 피드백 받거나 필요시 수정하면서 진행할 생각으로 커밋 하긴 했습니다.

개인적으로 컨트롤러가 어느 정도 기능적으로 존재는 해야 unit Test 든 통합 테스트 든 test가 의미가 있을 것 같습니다.

사실 아직도,
컨트롤러 부터 만들기 시작
1.API Spec을 바탕으로 DTO를 만들고,
2.Integration Test + Unit Test

부분이 이해되지 않는 부분이 많습니다.
그리고 어디서 부터 여쭤봐야할지도 모르겠습니다.
궁금한 점이나, 제대로 완성된 컨트롤러, DTO를 만들기 위해선
큰 순서가 있고, 현재 서로 개발시에 의존성이 있기도 하고해서 그런 것 같습니다.

제 생각이 틀리거나 잘못 이해한 것이 있어서 이렇게 헤매고 있을까요??

xenex90 replied:
그리고 분명 비즈니스 로직은 대부분의 경우에 DB에 저장하는 경우를 거치는데
service나 entity, dao등은 구현되지 않아서 하나의 비지니스 로직이 현재 멘토님의 말씀으로는
완성되기 어렵습니다. 제가 추가적으로 따로 진행해야 하는 부분일까요??

비즈니스 로직이 완성이 되면 그제서야 unit test를 해볼 수 있다고 생각합니다.
제가 unit test를 다시 한번 더 알아봐야 하는데, 제 현제까지의 지식으로는 그렇습니다.

혹시나 싶어서 멘토님께 여쭤봅니다.

xenex90 replied:
> 이것 대신에 @SpringBootTest(...RANDOM_PORT)를 써서 테스트를 작성해 보시면 좋을 것 같아요. 그리고, 그 두개는 테스트의 목적에서 어떤 차이가 있을까요?

한번 이것 반영해서 커밋 해보겠습니다.

f-lab-lyan commented on user-service/src/test/java/flab/userservice/UserControllerUnitTest.java:
```diff
@@ -0,0 +1,64 @@
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+@WebMvcTest(UserController.class)
+class UserControllerUnitTest {
```
컨트롤러 유닛 테스트와 컨트롤러 통합 테스트는 어떤 차이가 있는 것인가요? 목적 및 목표 관점에서 생각해보시면 좋을 것 같습니다.

xenex90 replied:
컨트롤러 유닛 테스트
하나의 로직 혹은 부분이 동작하는 지 확인 하는 테스트
컨트롤러 통합 테스트
컨트롤러 전체의 로직 혹은 컨트롤러가 온전히 기능하고 있는지 확인하는 테스트


목적과 목표의 관점
목적 관점
유닛 테스트
-> 개발 과정 중 혹은 특정한 조건에 부합하는 기능인지 확인 하는 테스트
통합 테스트
-> 출시 전에 로직이나 기능이 서로 잘 연동되어서 컨트롤러가 완성되었는지 확인하는 테스트

목표 관점
유닛, 통합 테스트
어플리케이션 내에서 컨트롤러가 처음에 설정한 목적에 맞게 잘 개발 되었는지 확인하는 테스트

라고 생각합니다.

xenex90 committed:
UserController modified, ApiResponse, UserLoginResquest, UserREsponse, UserSignupRequest add

-UserController http상태 코드 반영(201,409,500)
-기존 DTO를 회원가입, 로그인, 응답 경우로 분리

xenex90 committed:
-유닛 테스트 추가 및 user 컨트롤러 로그 기능 추가, 기타 코드 정리

-UserController,UserControllerUnitTest, build.gradle.kts, etc

xenex90 commented on common/src/main/java/flab/common/api/ApiResponse.java:
```diff
@@ -0,0 +1,34 @@
+package flab.common.api;
+
+
+public class ApiResponse {
```
ApiResponse을 common 모듈에 추가해서 응답에 사용하려고 합니다.

xenex90 commented on user-service/build.gradle.kts:
```diff
@@ -7,6 +7,7 @@ plugins {
 
 dependencies {
     implementation(project(":common"))
+    implementation("org.springframework.boot:spring-boot-starter-validation:3.2.5")
```
@NotBlank 어노테이션 관련 의존성 추가하였습니다.

f-lab-lyan replied:
이건 버전을 빼도 되지 않을까요?

xenex90 committed:
-user 컨트롤러 코드 정리(주석제거)

xenex90 commented on user-service/src/main/java/flab/userservice/dto/UserResponse.java:
```diff
@@ -0,0 +1,17 @@
+
+import java.time.LocalDate;
+
+public record UserResponse(
```
UserResponse클래스로 유저 응답에 쓸 DTO 입니다.

xenex90 commented on user-service/src/main/java/flab/userservice/controller/UserController.java:
```diff
@@ -0,0 +1,59 @@
+    public ApiResponse signup(@RequestBody UserSignupRequest request) {
+
+        try{
+            if(
```
비즈니스 로직(서비스)가 있다고 가정하고
성공 , 실패, 에러의 경우로 나누었습니다.

xenex90 replied:
컨트롤러를 더 개발하려고 했으나, 여기서 막혀 있습니다.

xenex90 commented on user-service/src/test/java/flab/userservice/UserControllerUnitTest.java:
```diff
@@ -0,0 +1,77 @@
+    void signupCallsService() throws Exception {
+        // 서비스가 정상적으로 처리했다고 가정
+        UserSignupRequest request = new UserSignupRequest(
+                "john_doe",
```
회원가입 성공의 경우의 값으로 test 하였습니다.(Test 결과는 성공)

xenex90 commented on user-service/src/test/java/flab/userservice/UserControllerUnitTest.java:
```diff
@@ -0,0 +1,77 @@
+    @DisplayName("로그인 요청이 들어오면 서비스 호출 후 200 반환")
+    void loginCallsService() throws Exception {
+        // 서비스가 토큰 객체를 반환한다고 가정
+       /* Mockito.when(userService.login(any(UserLoginRequest.class)))
```
비즈니스로직이 없어서
실패나 에러의 경우는 가정하기 힘들어서 일단 이대로 주석 처리 하였습니다.(비즈니스 로직)이 없는데도 테스트 할 수 있는지는 몰라서 한번 고민이나 알아보겠습니다.

xenex90 commented on user-service/src/main/java/flab/userservice/dto/UserSignupRequest.java:
```diff
@@ -0,0 +1,38 @@
+        @Size(max = 100)
+        String address,
+
+        @Pattern(regexp = "^\\d{4}-\\d{2}-\\d{2}$", message = "날짜 형식은 yyyy-MM-dd 여야 합니다.")
```
birthDate의 경우 String 타입으로 변경하고 조건을 거는 것으로 충분히 validation이 가능하다고 보고 LocalDate -> String 타입으로 변경하였습니다.

xenex90 committed:
-UserControllerRandomPortUnitTest 추가

RandomPort Spring Test 경우를 추가하고
컨트롤러에서 Header 부분 추가하고 UnitTest에서 header에서 검사하도록 수정하였습니다.

xenex90 commented on user-service/src/main/java/flab/userservice/controller/UserController.java:
```diff
@@ -0,0 +1,78 @@
+            ){
+                //성공시
+                log.info("회원가입 성공");
+                HttpHeaders headers = new HttpHeaders();
```
httpheaders를 추가해서 header로 response를 return 하도록 하였습니다.

xenex90 commented on user-service/src/test/java/flab/userservice/controller/UserControllerRandomPortUnitTest.java:
```diff
@@ -0,0 +1,60 @@
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.jsonPath;
+import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;
+
+@SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
```
말씀하신 SpringBootTest RANDOM_PORT 옵션을 사용해서 UnitTest를 개발하여 보았습니다.

xenex90 commented on user-service/src/test/java/flab/userservice/controller/UserControllerRandomPortUnitTest.java:
```diff
@@ -0,0 +1,60 @@
+
+        assertEquals(HttpStatus.CREATED, response.getStatusCode());
+        assertNotNull(response.getHeaders());
+        assertEquals("User registration success", response.getHeaders().get("Message").get(0));
```
아직 좀 엉성하긴 하지만 응답의 body에 status와 message를 넣는 것보다, 실패하더라도 header에 넣어서 처리하는 것이 낫다고 보고 미리 옮겼습니다.

xenex90 reviewed (commented):
2025.7월 9일분까지 변경사항 코드 확인하였습니다.

xenex90 committed:
-회원 가입 로직 개발 중, unitTest수행 중

로직 순서
1. 요청 중 해당 이메일이 DB에 존재하는지 확인
2. 존재하면 save메서드 수행(201 return)
3. 실패하거나 에러 발생하면 207 or 500 에러 return

xenex90 commented on user-service/build.gradle.kts:
```diff
@@ -8,6 +8,8 @@ plugins {
 dependencies {
     implementation(project(":common"))
     implementation("org.springframework.boot:spring-boot-starter-validation:3.2.5")
+    implementation ("org.springframework.boot:spring-boot-starter-data-jpa")
```
JPA 사용할 수 있는 의존성 추가 하였습니다.

xenex90 commented on user-service/src/main/java/flab/userservice/controller/UserController.java:
```diff
@@ -20,34 +22,36 @@
 @RequestMapping("/users")
 public class UserController {
 
+    @Autowired
```
의존성 주입 이렇게 하지 말라고 하던데
아직 이렇게 해도 되지 않나요??

xenex90 commented on user-service/src/main/java/flab/userservice/controller/UserController.java:
```diff
@@ -20,34 +22,36 @@
-                return new ResponseEntity<>(response, headers, HttpStatus.CONFLICT);
-            }
+
+                if(userService.existsByEmail(request.email())){
```
email 존재하는 지 확인하는 서비스 호출 하고 분기합니다.

xenex90 commented on user-service/src/main/java/flab/userservice/entity/User.java:
```diff
@@ -0,0 +1,51 @@
+@NoArgsConstructor
+@AllArgsConstructor
+@Builder
+public class User {
```
entity 추가하였습니다.

xenex90 commented on user-service/src/main/java/flab/userservice/repository/UserRepository.java:
```diff
@@ -0,0 +1,33 @@
+
+import java.util.Optional;
+
+@Repository
```
Repository 추가하였습니다.

대강 만들어서 추가하였습니다.
추후 수정을 더 진행하겠습니다.

xenex90 commented on user-service/src/main/java/flab/userservice/service/UserService.java:
```diff
@@ -0,0 +1,52 @@
+
+@Service
+@RequiredArgsConstructor
+public class UserService {
```
서비스 추가하였습니다.

아직 완성은 아니라 추후 더 수정진행하겠습니다.

xenex90 commented on user-service/src/main/resources/application.yml:
```diff
@@ -0,0 +1,15 @@
+spring:
+  datasource:
```
로컬 db 설정 파일입니다.

xenex90 reviewed (commented):
간단히 코멘트 달았습니다.
확인 부탁드립니다.

f-lab-lyan commented on user-service/src/main/java/flab/userservice/controller/UserController.java:
```diff
@@ -0,0 +1,82 @@
+@RequestMapping("/users")
+public class UserController {
+
+    @Autowired
+    private UserService userService;
```
```suggestion
    UserController userController = new UserController();
    UserService userService = new UserService();
```

f-lab-lyan commented on user-service/src/main/java/flab/userservice/service/UserService.java:
```diff
@@ -0,0 +1,52 @@
+import java.util.Optional;
+
+@Service
+@RequiredArgsConstructor
```
```suggestion
    UserRepository userRepository = new UserRepository();
    UserService userService = new UserService(userRepository);
```

f-lab-lyan commented on user-service/src/main/java/flab/userservice/controller/UserController.java:
```diff
@@ -0,0 +1,82 @@
+    public ResponseEntity<Object> signup(@RequestBody UserSignupRequest request) {
+
+
+        try{
```
ExceptionHandler를 사용하는 것이 좋지 않을까요?

f-lab-lyan commented on user-service/src/main/java/flab/userservice/controller/UserController.java:
```diff
@@ -0,0 +1,82 @@
+            HttpHeaders headers = new HttpHeaders();
+            ApiResponse response = new ApiResponse("Server Internal Error", 500);
+            headers.add("Message", response.getMessage());
+            headers.add("Status", String.valueOf(response.getStatus()));
```
ResponseEntity<?>

f-lab-lyan reviewed (approved):
수고하셨습니다. 👍
